depends: build zeromq with CMake

fanquake opened the pull request:
This picks up a change, which is a switch to building zeromq with CMake. It includes a number of patches, some which have already been upstreamed (see each patch for details).

theuni committed:
depends: switch zmq to CMake

The CMake WIN32_WINNT autodetection is broken, and must be set
manually. We may want to set is explicitly in any case, but the
brokenness should also be fixed upstream.

Also patch out depends paths, that would cause non-determinism.

Co-authored-by: fanquake <fanquake@gmail.com>

hebasto commented:
Concept ACK.

hebasto commented:
> This picks up a change, which is a switch to building zeromq with CMake.

From https://github.com/zeromq/libzmq/pull/4667#issue-2197755398:
> Please note that CMake-generated `libzmq.pc` file is also broken as its "Libs.private" section contains only `-lstdc++` when cross-compiling for Windows.

fanquake commented:
> From https://github.com/zeromq/libzmq/pull/4667#issue-2197755398:

Can you elaborate / suggest something concrete? As far as I can see, the cross-compiling build of Windows for this branch currently works fine, we already link again `-lws2_32` from our own build (so that shouldn't cause build failures), and we currently patch the `-lstdc++` out of `Libs.private`, because as far as we are concerned, it's a bug.

fanquake commented:
Given both changes have landed, I've reordered the commits, and undrafted. Will followup with the Windows issues.

hebasto reviewed (commented):
I've tested 796a271f0f4b57f61f357aac3e7a76072fed6f9a on Ubuntu 23.10 using a patch from #29960.

There are a few differences between Autotools and CMake builds:

1. In CMake, the resulted archive lacks object files from the following sources:
```
gssapi_client.cpp
gssapi_mechanism_base.cpp
gssapi_server.cpp
vmci_address.cpp
vmci_connecter.cpp
vmci_listener.cpp
vmci.cpp
```
2. CMake build is effectively compiled with `-O3` flag.

3. CMake adds `-DZMQ_CUSTOM_PLATFORM_HPP`.

fanquake commented:
Fixed the Windows build error, but drafted while based on #30078.

fanquake commented:
> ld64.lld: error: library not found for -lrt

Still at least one other bugfix to upstream here. ZMQ is trying to link against realtime for macOS.

fanquake commented:
>  ZMQ is trying to link against realtime on macOS.

Patched out all usage of `librt` here. Landed one related change upstream: https://github.com/zeromq/libzmq/pull/4702.

fanquake commented:
Guix Build (aarch64):
```bash
4f3beefc6f4dc2a44829697ebd14e2f1016a35372f261bd379c66190e93db745  guix-build-f684b5d7ddfe/output/aarch64-linux-gnu/SHA256SUMS.part
b80ef38b4ad2a51bd033114884cb4b30e6b10675a6bfca08aa96680a7a754fc9  guix-build-f684b5d7ddfe/output/aarch64-linux-gnu/bitcoin-f684b5d7ddfe-aarch64-linux-gnu-debug.tar.gz
1c9c9198bda3419e9d49c809ef2a7e4e1c2b7846f9e03695e5175dbc7a5c0c8f  guix-build-f684b5d7ddfe/output/aarch64-linux-gnu/bitcoin-f684b5d7ddfe-aarch64-linux-gnu.tar.gz
3a9b4be095aa58dd7e99494d863d4d247fff110576f78e0682d6893868543b64  guix-build-f684b5d7ddfe/output/arm-linux-gnueabihf/SHA256SUMS.part
f6a0df95ce573affb505daf30b068f7baea6fdf6f489ee8cd88dd22c3f8e7f81  guix-build-f684b5d7ddfe/output/arm-linux-gnueabihf/bitcoin-f684b5d7ddfe-arm-linux-gnueabihf-debug.tar.gz
678f17eba53df3964445a8968029d41bbe6ab199eb7ce8b62ea96233bf1d6c54  guix-build-f684b5d7ddfe/output/arm-linux-gnueabihf/bitcoin-f684b5d7ddfe-arm-linux-gnueabihf.tar.gz
911e56f6d3f0e3b5d6003bd6b6c9c60c55976d03813af13bac54bffd6e282f0b  guix-build-f684b5d7ddfe/output/arm64-apple-darwin/SHA256SUMS.part
4a07c8e8ffe0fb0be25bd26bd0c3e6e3b921de6eb083625bd3d5f071520f7364  guix-build-f684b5d7ddfe/output/arm64-apple-darwin/bitcoin-f684b5d7ddfe-arm64-apple-darwin-unsigned.tar.gz
bb1c090b2ce12232ce7934a6f386bd23b0e14f69f0bd8e7b2adfbbca689cee15  guix-build-f684b5d7ddfe/output/arm64-apple-darwin/bitcoin-f684b5d7ddfe-arm64-apple-darwin-unsigned.zip
e4304bbbd976af67ff6ba1dc1abbbc4dd0ad1dae42a6822f2cf2548b9fd8f980  guix-build-f684b5d7ddfe/output/arm64-apple-darwin/bitcoin-f684b5d7ddfe-arm64-apple-darwin.tar.gz
b5732915ddfebcda4b2ec4fa5f7b5fdd88783bccd5729c1d3a64ba207806ce60  guix-build-f684b5d7ddfe/output/dist-archive/bitcoin-f684b5d7ddfe.tar.gz
ef255118bf5f06508eba468c92fd0f33e1edb55db00d605ebdff2aa509aa52cc  guix-build-f684b5d7ddfe/output/powerpc64-linux-gnu/SHA256SUMS.part
1fb84edc34acc73ce32017f5b916468fd86595ae36b7e952cdc5848d73e6dbea  guix-build-f684b5d7ddfe/output/powerpc64-linux-gnu/bitcoin-f684b5d7ddfe-powerpc64-linux-gnu-debug.tar.gz
ad0f5ca76dff888294e21c83e2f492b5cf347000608280865fb9e7d236f12c45  guix-build-f684b5d7ddfe/output/powerpc64-linux-gnu/bitcoin-f684b5d7ddfe-powerpc64-linux-gnu.tar.gz
868d1f43e4d1560b24349d4e1d92b91520a731ef6de15389b0747f9012ce4cf0  guix-build-f684b5d7ddfe/output/riscv64-linux-gnu/SHA256SUMS.part
0f5e47a6aaa673a2fb4ad7b0e21c6f732ad918d8fc567ca0e8f1ace25299ece0  guix-build-f684b5d7ddfe/output/riscv64-linux-gnu/bitcoin-f684b5d7ddfe-riscv64-linux-gnu-debug.tar.gz
c24e814710fac82e923ebc930c4eb63f0e7aaa58ef5112934bbac7570f6e28f2  guix-build-f684b5d7ddfe/output/riscv64-linux-gnu/bitcoin-f684b5d7ddfe-riscv64-linux-gnu.tar.gz
74065401825da30ea6cadaa3e1df6ce3dc9549fb27720328b39a4a18493baedd  guix-build-f684b5d7ddfe/output/x86_64-apple-darwin/SHA256SUMS.part
c2c1224bea758b01eb1437c15e521d54eb156be456ff049fc640f6985fd1c8a5  guix-build-f684b5d7ddfe/output/x86_64-apple-darwin/bitcoin-f684b5d7ddfe-x86_64-apple-darwin-unsigned.tar.gz
829700ec45a0dafc85150ea8d625d1c7b314a6db46003819996385f5df36745d  guix-build-f684b5d7ddfe/output/x86_64-apple-darwin/bitcoin-f684b5d7ddfe-x86_64-apple-darwin-unsigned.zip
ff9939686a5aae5d9720817d2f60764625e2167fb25f55f74f726961df588228  guix-build-f684b5d7ddfe/output/x86_64-apple-darwin/bitcoin-f684b5d7ddfe-x86_64-apple-darwin.tar.gz
60917fe01cd681e8d85e4fdb3786e77e3aa5675de064862ba44f7d7ed74440c3  guix-build-f684b5d7ddfe/output/x86_64-linux-gnu/SHA256SUMS.part
182777e922f62e88ca03ca34405c40e6961920d4da7e1d79bf0e96131bfafa97  guix-build-f684b5d7ddfe/output/x86_64-linux-gnu/bitcoin-f684b5d7ddfe-x86_64-linux-gnu-debug.tar.gz
5477f3dd03e2866d16d73aa62eae83cabe27622aaefc05d7cbc87fa4a3f4ed88  guix-build-f684b5d7ddfe/output/x86_64-linux-gnu/bitcoin-f684b5d7ddfe-x86_64-linux-gnu.tar.gz
f3ede7e34d810ba19b1455e20b7f8161cdae38945edb6a60cee90f083ae86e81  guix-build-f684b5d7ddfe/output/x86_64-w64-mingw32/SHA256SUMS.part
2fd9d41e41787edbfa8572d956b957edc0ad9134cae60d6cfff4d50b8cf099bb  guix-build-f684b5d7ddfe/output/x86_64-w64-mingw32/bitcoin-f684b5d7ddfe-win64-debug.zip
600b367d64e6595d06e0409f9d6b4e33c5277978bc64092412afc094411b18a5  guix-build-f684b5d7ddfe/output/x86_64-w64-mingw32/bitcoin-f684b5d7ddfe-win64-setup-unsigned.exe
ada26ab5b9d21238e0babf3fc49a3dadd7d9d07828fd5fb996ca35c99aae1770  guix-build-f684b5d7ddfe/output/x86_64-w64-mingw32/bitcoin-f684b5d7ddfe-win64-unsigned.tar.gz
2e80dfbce201a843fb68b68ef01600237297eeed03f475a21826013eb70803dc  guix-build-f684b5d7ddfe/output/x86_64-w64-mingw32/bitcoin-f684b5d7ddfe-win64.zip
```

fanquake committed:
depends: add zeromq mktemp macos patch

fanquake committed:
depends: add zeromq builtin sha1 patch

fanquake committed:
depends: add zeromq windows usage patch

fanquake committed:
depends: add zeromq cmake minimum patch

fanquake committed:
depends: add zeromq no librt patch

fanquake commented:
Rebased, and reworked the changes here, to hopefully make them easier to review.

hebasto commented:
> > From [zeromq/libzmq#4667 (comment)](https://github.com/zeromq/libzmq/pull/4667#issue-2197755398):
> 
> Can you elaborate / suggest something concrete?

Please see https://github.com/zeromq/libzmq/pull/4706.

hebasto commented on depends/patches/zeromq/fix_have_windows.patch:
```diff
@@ -0,0 +1,62 @@
+commit b163fc36a48ecdc87a3ecb4c6bba5f6b8965acaf
```
Cannot find b163fc36a48ecdc87a3ecb4c6bba5f6b8965acaf in https://github.com/zeromq/libzmq. Did you mean https://github.com/zeromq/libzmq/commit/aa885c5a154256612108636b0fb22f44ae0e247a?

Perhaps we could refer to a PR number, as is done in other patches?

hebasto commented on depends/patches/zeromq/cmake_minimum.patch:
```diff
@@ -0,0 +1,18 @@
++++ b/CMakeLists.txt
+@@ -1,12 +1,7 @@
+ # CMake build script for ZeroMQ
++cmake_minimum_required(VERSION 3.16)
```
Unless this patch is submitted upstream, why not use the same version as our CMake-based build system?

fanquake replied:
Reverting this, otherwise this will just be blocked by the CI (without other changes that I'm not making here), and matching the CMake version isn't a requirement.

hebasto commented on depends/patches/zeromq/no_librt.patch:
```diff
@@ -0,0 +1,54 @@
+index 03462271..87ceab3c 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -569,13 +569,6 @@ else()
```
```suggestion
@@ -564,13 +564,6 @@ else()
```

hebasto commented on depends/patches/zeromq/no_librt.patch:
```diff
@@ -0,0 +1,54 @@
+ find_package(Threads)
+ 
+ if(WIN32 AND NOT CYGWIN)
+@@ -593,9 +586,7 @@ if(WIN32 AND NOT CYGWIN)
```
```suggestion
@@ -588,9 +581,7 @@ if(WIN32 AND NOT CYGWIN)
```

hebasto commented on depends/patches/zeromq/no_librt.patch:
```diff
@@ -0,0 +1,54 @@
+ 
+   check_cxx_symbol_exists(fork unistd.h HAVE_FORK)
+   check_cxx_symbol_exists(gethrtime sys/time.h HAVE_GETHRTIME)
+@@ -1508,10 +1499,6 @@ if(BUILD_SHARED)
```
```suggestion
@@ -1503,10 +1494,6 @@ if(BUILD_SHARED)
```

hebasto commented on depends/patches/zeromq/no_librt.patch:
```diff
@@ -0,0 +1,54 @@
+   if(norm_FOUND)
+       target_link_libraries(libzmq norm::norm)
+   endif()
+@@ -1558,10 +1545,6 @@ if(BUILD_STATIC)
```
```suggestion
@@ -1553,10 +1540,6 @@ if(BUILD_STATIC)
```

hebasto reviewed (commented):
Approach ACK f2712ed278ec44ad100bdbe765de1716191f238b.

I've reviewed all patches for now.

fanquake commented:
Rebased. Added all the missing newline/diff adjustments. Swapped CMake minimum patch to 3.22. Updated the PR reference in the Windows patch.

hebasto reviewed (commented):
Tested 22405f6f777e8701cbef0103fcfa06fd9bd109cb.

By default, the `ZeroMQ` project sets `CMAKE_BUILD_TYPE` to `Release`, which implies the `-O3 -DNDEBUG` flags override our `$(host)_release_CXXFLAGS`.

I suggest passing `-DCMAKE_BUILD_TYPE=None` to disable this overriding, which also works with `DEBUG=1`.

hebasto commented on depends/packages/zeromq.mk:
```diff
@@ -3,32 +3,42 @@ $(package)_version=4.3.5
+  $(package)_config_opts += -DBUILD_SHARED=OFF -DBUILD_TESTS=OFF -DZMQ_BUILD_TESTS=OFF
+  $(package)_config_opts += -DENABLE_DRAFTS=OFF
+  $(package)_config_opts_mingw32 += -DZMQ_WIN32_WINNT=0x0601 -DZMQ_HAVE_IPC=OFF
+  $(package)_config_opts_debug := -DWITH_PERF_TOOL=OFF
```
If I'm not mistaken, we currently have no means to convert `DEBUG=1` into `-DCMAKE_BUILD_TYPE=Debug`, which means that this line will never be used.

fanquake commented:
> I suggest passing -DCMAKE_BUILD_TYPE=None to disable this overriding, which also works with DEBUG=1.

I'm not exactly sure what we want to do with build types and CMake packages in depends going forward, but have used `None` here for now.

hebasto commented:
My Guix build:
```
x86_64
0d9925c3463e882ead8b87010ef9ce477107cdb3d1da54c9e50a70ed7c128463  guix-build-218e69c705d7/output/aarch64-linux-gnu/SHA256SUMS.part
0fea765c772ee00ec37d4efde6296e173a8a765792534a731e64b253effd9242  guix-build-218e69c705d7/output/aarch64-linux-gnu/bitcoin-218e69c705d7-aarch64-linux-gnu-debug.tar.gz
4a2c51df14f82f5947f12ec28274d6b65f42e7e03f766d1a52937b7fa4075bbf  guix-build-218e69c705d7/output/aarch64-linux-gnu/bitcoin-218e69c705d7-aarch64-linux-gnu.tar.gz
f061bd6ffb398fc95b71515794f2bf8f32a401d8a4d6f41ed0bdf4c62ad22730  guix-build-218e69c705d7/output/arm-linux-gnueabihf/SHA256SUMS.part
6b205c30827be076573277d6741b7f798c4c66f6ab5f596416e00fc3561c3782  guix-build-218e69c705d7/output/arm-linux-gnueabihf/bitcoin-218e69c705d7-arm-linux-gnueabihf-debug.tar.gz
3eee660dd3f26e530ec5742d3939461b60c19d7d803cf3564bf9d9d6ee60e7ff  guix-build-218e69c705d7/output/arm-linux-gnueabihf/bitcoin-218e69c705d7-arm-linux-gnueabihf.tar.gz
2e4fde768d4157dded1f1f5892e2020bcb606d801aedaabc736880d93f66eb82  guix-build-218e69c705d7/output/arm64-apple-darwin/SHA256SUMS.part
d85a9cb0e1d16257c56f646ca2982a7b2ec8c392ee1a5795077022afd1aef351  guix-build-218e69c705d7/output/arm64-apple-darwin/bitcoin-218e69c705d7-arm64-apple-darwin-unsigned.tar.gz
35535250213e661c7a1a027cea8939f1f99727fa24d4ef419f00a3fd20428595  guix-build-218e69c705d7/output/arm64-apple-darwin/bitcoin-218e69c705d7-arm64-apple-darwin-unsigned.zip
33c06b7d2cf16f3f04e5692ad07ac8e945729b76daae24304ecc1507b258d0a5  guix-build-218e69c705d7/output/arm64-apple-darwin/bitcoin-218e69c705d7-arm64-apple-darwin.tar.gz
22cf0c65544a061587581caa67bb5d4a110c413ba4dc49bde9d65c03d045379a  guix-build-218e69c705d7/output/dist-archive/bitcoin-218e69c705d7.tar.gz
e17664cc9bddf356405b339bd2ee0aba78c37d10c90dfb44b35f8f5f50a3627a  guix-build-218e69c705d7/output/powerpc64-linux-gnu/SHA256SUMS.part
efbba26bac27c0e84892fab7a6a02583fe2caa346c89999b2b35a39ef88f285c  guix-build-218e69c705d7/output/powerpc64-linux-gnu/bitcoin-218e69c705d7-powerpc64-linux-gnu-debug.tar.gz
b76da73f6f4b35484f85f34874f0f877ea2a2c6048741ccbab4af72c72125c22  guix-build-218e69c705d7/output/powerpc64-linux-gnu/bitcoin-218e69c705d7-powerpc64-linux-gnu.tar.gz
48af076d3ba27e5e5b40df217c38397a845040b6d98f683ed011644c6ff61a49  guix-build-218e69c705d7/output/riscv64-linux-gnu/SHA256SUMS.part
35bd7a2914158357b72422bb7105ac4113c933fbf7f3c29063f0db38355ef631  guix-build-218e69c705d7/output/riscv64-linux-gnu/bitcoin-218e69c705d7-riscv64-linux-gnu-debug.tar.gz
d2adaee5944838a83ec9f2633a4ea355b25a310d34de22ec250c676b335206f2  guix-build-218e69c705d7/output/riscv64-linux-gnu/bitcoin-218e69c705d7-riscv64-linux-gnu.tar.gz
6395aa98b9af78eac690b2d9c70ca9c58ca29c735263a9ceee3c7ce733fe6096  guix-build-218e69c705d7/output/x86_64-apple-darwin/SHA256SUMS.part
a44a80dbfded848aeeccd8b91bff2ebaf74bbfd0ce1d8bab02a142a890456a85  guix-build-218e69c705d7/output/x86_64-apple-darwin/bitcoin-218e69c705d7-x86_64-apple-darwin-unsigned.tar.gz
89fd027f38577c04c98fdd5d2bf0a732ae801f601846323c9887d3d9e7ee91d6  guix-build-218e69c705d7/output/x86_64-apple-darwin/bitcoin-218e69c705d7-x86_64-apple-darwin-unsigned.zip
76c017c1d7fbf872e810ce7315acfa3212292f42cf12c8ad60d106eb5a8b423f  guix-build-218e69c705d7/output/x86_64-apple-darwin/bitcoin-218e69c705d7-x86_64-apple-darwin.tar.gz
534d3cbbae39f0b57a91bc413c6f6b33b7dd67a440b551b6851af86c8643fad6  guix-build-218e69c705d7/output/x86_64-linux-gnu/SHA256SUMS.part
4bedcfe7bbfc9047b220ed5a099a5857ee78dd3c58bf027fa859d6a8af474162  guix-build-218e69c705d7/output/x86_64-linux-gnu/bitcoin-218e69c705d7-x86_64-linux-gnu-debug.tar.gz
b660a9155a52b0f8d81250253138d8cbf7ec565ac5687caaa60244166ea15fcd  guix-build-218e69c705d7/output/x86_64-linux-gnu/bitcoin-218e69c705d7-x86_64-linux-gnu.tar.gz
9c4744ce005a38fae27252282c74b89348ea982bbf972741dc1952f3e6731d6f  guix-build-218e69c705d7/output/x86_64-w64-mingw32/SHA256SUMS.part
7758caae1f39ac570b28f72339ad6ecdc493ba424c027c910f2e98154a34be71  guix-build-218e69c705d7/output/x86_64-w64-mingw32/bitcoin-218e69c705d7-win64-debug.zip
1ef0eca4b85e8390887b053491f3695849d5804f9db3bad8dde75b1191d8f5cf  guix-build-218e69c705d7/output/x86_64-w64-mingw32/bitcoin-218e69c705d7-win64-setup-unsigned.exe
9f65f38cf2bb7e2883cd02c209f21fd817ab68803c6cc123c4b37010ea2010f9  guix-build-218e69c705d7/output/x86_64-w64-mingw32/bitcoin-218e69c705d7-win64-unsigned.tar.gz
4a9dda1fd06c54ce9d663b8a1625665e9dd2c40abf398b83d66a4bc9d7978ae2  guix-build-218e69c705d7/output/x86_64-w64-mingw32/bitcoin-218e69c705d7-win64.zip
```

fanquake commented:
Guix Build (aarch64):
```bash
e9a70ec7c66652aa96facc2fe3117fb177cb61e6d28fadc48893546f9cebf6cb  guix-build-218e69c705d7/output/aarch64-linux-gnu/SHA256SUMS.part
0fea765c772ee00ec37d4efde6296e173a8a765792534a731e64b253effd9242  guix-build-218e69c705d7/output/aarch64-linux-gnu/bitcoin-218e69c705d7-aarch64-linux-gnu-debug.tar.gz
1689943e2b3caa664a9bb5c137ffc161252b253568b563d5a4433e2ac55174ec  guix-build-218e69c705d7/output/aarch64-linux-gnu/bitcoin-218e69c705d7-aarch64-linux-gnu.tar.gz
6934a443a689d40d61fd22ed99596971e3d0d72dcb547f45e9298650410073a4  guix-build-218e69c705d7/output/arm-linux-gnueabihf/SHA256SUMS.part
6b205c30827be076573277d6741b7f798c4c66f6ab5f596416e00fc3561c3782  guix-build-218e69c705d7/output/arm-linux-gnueabihf/bitcoin-218e69c705d7-arm-linux-gnueabihf-debug.tar.gz
d9e54b54890aab83a4a197f319b369c97867630040e0836a94af0d55a9965c73  guix-build-218e69c705d7/output/arm-linux-gnueabihf/bitcoin-218e69c705d7-arm-linux-gnueabihf.tar.gz
0ece588fdc72a3ba98ef6b95a0327a7ed40cf75128b121333fd0ee4b251abde8  guix-build-218e69c705d7/output/arm64-apple-darwin/SHA256SUMS.part
abaeac99b9170c43ce49931b61adecabe6cc5278b1c9598baaf3bff6c6f18732  guix-build-218e69c705d7/output/arm64-apple-darwin/bitcoin-218e69c705d7-arm64-apple-darwin-unsigned.tar.gz
e17bab169418a3ee3b1f3679a5b597de62b40744e13d845171c576760b1ca61c  guix-build-218e69c705d7/output/arm64-apple-darwin/bitcoin-218e69c705d7-arm64-apple-darwin-unsigned.zip
a92ccb4fdd1b3f89a61e890fe5670ecb2b0553841489a09c30c37c5654f236e7  guix-build-218e69c705d7/output/arm64-apple-darwin/bitcoin-218e69c705d7-arm64-apple-darwin.tar.gz
22cf0c65544a061587581caa67bb5d4a110c413ba4dc49bde9d65c03d045379a  guix-build-218e69c705d7/output/dist-archive/bitcoin-218e69c705d7.tar.gz
1ce094cc2a24c62e9731af914f53d72c699e32cc43acb5e686bb01328e4ea5ac  guix-build-218e69c705d7/output/powerpc64-linux-gnu/SHA256SUMS.part
efbba26bac27c0e84892fab7a6a02583fe2caa346c89999b2b35a39ef88f285c  guix-build-218e69c705d7/output/powerpc64-linux-gnu/bitcoin-218e69c705d7-powerpc64-linux-gnu-debug.tar.gz
b56a788dda4e2a312be7beeda3a66e47877053cfb6cd17d7a0f4586529ea70d3  guix-build-218e69c705d7/output/powerpc64-linux-gnu/bitcoin-218e69c705d7-powerpc64-linux-gnu.tar.gz
4f5bdea02bcff7900e52ea6cd27f8a942669505033beabee28d54f1c0d63f8dc  guix-build-218e69c705d7/output/riscv64-linux-gnu/SHA256SUMS.part
35bd7a2914158357b72422bb7105ac4113c933fbf7f3c29063f0db38355ef631  guix-build-218e69c705d7/output/riscv64-linux-gnu/bitcoin-218e69c705d7-riscv64-linux-gnu-debug.tar.gz
4bc8cf3137e5148ab76d273801dacc65974acec47de5f4c1d1771cbe3a290aa8  guix-build-218e69c705d7/output/riscv64-linux-gnu/bitcoin-218e69c705d7-riscv64-linux-gnu.tar.gz
25547763836679e87c575f85676afee3237bdfb0fca94ed148ddbbda4c0805e4  guix-build-218e69c705d7/output/x86_64-apple-darwin/SHA256SUMS.part
90b561e0db776fd5ac7bda02c9b4d00c6ff8967358ea7fd3eb2ff001ff6b875a  guix-build-218e69c705d7/output/x86_64-apple-darwin/bitcoin-218e69c705d7-x86_64-apple-darwin-unsigned.tar.gz
0e3fdc1ddee44712df23642226bd2b4945660eedb0472bc55d0f5596354713a1  guix-build-218e69c705d7/output/x86_64-apple-darwin/bitcoin-218e69c705d7-x86_64-apple-darwin-unsigned.zip
9ead4515b6c2eda88a211e78f18ba30aa7c29c9c323cf6406d19a2131bb3b136  guix-build-218e69c705d7/output/x86_64-apple-darwin/bitcoin-218e69c705d7-x86_64-apple-darwin.tar.gz
84c508f46fa3a840b39d9620c760a92d278b68b9f65a61176494a45a176bc981  guix-build-218e69c705d7/output/x86_64-linux-gnu/SHA256SUMS.part
4bedcfe7bbfc9047b220ed5a099a5857ee78dd3c58bf027fa859d6a8af474162  guix-build-218e69c705d7/output/x86_64-linux-gnu/bitcoin-218e69c705d7-x86_64-linux-gnu-debug.tar.gz
0d11ab9bc8fe9551d049e92b30e5723c2df7b7afb88dae4ea7a4f007e587ce04  guix-build-218e69c705d7/output/x86_64-linux-gnu/bitcoin-218e69c705d7-x86_64-linux-gnu.tar.gz
b0135845804186285531b453cf9476c2fcb75b9f99815d4c28922efdb86119c6  guix-build-218e69c705d7/output/x86_64-w64-mingw32/SHA256SUMS.part
7758caae1f39ac570b28f72339ad6ecdc493ba424c027c910f2e98154a34be71  guix-build-218e69c705d7/output/x86_64-w64-mingw32/bitcoin-218e69c705d7-win64-debug.zip
a741d5c0424593c903694c79c938c36c313097d49fee0306067467b60950bb43  guix-build-218e69c705d7/output/x86_64-w64-mingw32/bitcoin-218e69c705d7-win64-setup-unsigned.exe
e8580fbc2905c0ad12d1f2347d54ca21213be13839ed0679df26da43c1325bd1  guix-build-218e69c705d7/output/x86_64-w64-mingw32/bitcoin-218e69c705d7-win64-unsigned.tar.gz
a59342f3acc2ccedc95d55c27745f22e677720e317a9ed17c85a78fb8881b859  guix-build-218e69c705d7/output/x86_64-w64-mingw32/bitcoin-218e69c705d7-win64.zip
```

Given that these don't match the `x86_64` build above, it looks like zeromqs CMake build system is non-deterministic across architectures.

fanquake commented:
> is non-deterministic across architectures.

Should be fixed now.

hebasto commented:
My Guix build:
```
x86_64
5cc76ca8f123e7ff2a365d6a94246a949c4976cf9ecfb9056a4e696a1b80b097  guix-build-e096f5d456f7/output/aarch64-linux-gnu/SHA256SUMS.part
1c6c8d2efba8f6cbe54b8e8bde00d962f9837ab76f801507a370e4d15ada0059  guix-build-e096f5d456f7/output/aarch64-linux-gnu/bitcoin-e096f5d456f7-aarch64-linux-gnu-debug.tar.gz
27110e79b03a3f22c24888f79e84c684eaddfb04b8595d200d6db5d689f260a2  guix-build-e096f5d456f7/output/aarch64-linux-gnu/bitcoin-e096f5d456f7-aarch64-linux-gnu.tar.gz
ddb3c16f6a780e1e0d63d4b948e321d6e97f3e25e31af7ebd0244def56ae42f7  guix-build-e096f5d456f7/output/arm-linux-gnueabihf/SHA256SUMS.part
ef8c0e7340e0722df2b3a002cdb8fcc9e2bb340e5e304ea44a95d4f5e56041af  guix-build-e096f5d456f7/output/arm-linux-gnueabihf/bitcoin-e096f5d456f7-arm-linux-gnueabihf-debug.tar.gz
f61732606be8a71f7759b4a6a349e4df7d0853f1780a20f6567230c488c2095c  guix-build-e096f5d456f7/output/arm-linux-gnueabihf/bitcoin-e096f5d456f7-arm-linux-gnueabihf.tar.gz
ab5e98243b1c778918bed05981ca43ca10a792e3ce65a9989f0ee21ad9e6e2f3  guix-build-e096f5d456f7/output/arm64-apple-darwin/SHA256SUMS.part
37250fd9082e616388ae28eb545ea66d199c5d6dc20408654d29ae3ad843816e  guix-build-e096f5d456f7/output/arm64-apple-darwin/bitcoin-e096f5d456f7-arm64-apple-darwin-unsigned.tar.gz
51cc9248dab86d10fc098e73a532d30285cfcc69acb2149ea31e71e6acb8672f  guix-build-e096f5d456f7/output/arm64-apple-darwin/bitcoin-e096f5d456f7-arm64-apple-darwin-unsigned.zip
c2f38544f46a68f82da1eb557a20f6c9d01e21117a5cc7b6c87e33eeb63cb487  guix-build-e096f5d456f7/output/arm64-apple-darwin/bitcoin-e096f5d456f7-arm64-apple-darwin.tar.gz
3696e4a23ed4cf989b2cf0cf0bf63443a6b57a0293af9257ba0be27dda759f24  guix-build-e096f5d456f7/output/dist-archive/bitcoin-e096f5d456f7.tar.gz
1d107f44afa9ea55ebf2bbc260873c5dd5ffa47cd1a0926b4d77c8f5de7de951  guix-build-e096f5d456f7/output/powerpc64-linux-gnu/SHA256SUMS.part
799e62815d67d8573ad2c5a6afa5a887b296bf4540dcbcdcd52c7823668b589e  guix-build-e096f5d456f7/output/powerpc64-linux-gnu/bitcoin-e096f5d456f7-powerpc64-linux-gnu-debug.tar.gz
f4fcc32445793ac22262658173b781b86292a25bf4903e828ad41b415be5b165  guix-build-e096f5d456f7/output/powerpc64-linux-gnu/bitcoin-e096f5d456f7-powerpc64-linux-gnu.tar.gz
3a7e79df66970de9d78012a6d452646f7bfc10fc58cbfe339acda4830d644985  guix-build-e096f5d456f7/output/riscv64-linux-gnu/SHA256SUMS.part
720ec3c0ac533271dcc34a5cc51b0ce54e0c03aefcdf60dca013d7350ac55850  guix-build-e096f5d456f7/output/riscv64-linux-gnu/bitcoin-e096f5d456f7-riscv64-linux-gnu-debug.tar.gz
02453db4bac48041702f26dc344b6a46bc5d28c97db594d7726a27309242d866  guix-build-e096f5d456f7/output/riscv64-linux-gnu/bitcoin-e096f5d456f7-riscv64-linux-gnu.tar.gz
e6a79a30ba94d4952a6b8d19b992418c6ebf0d9aef7b257ef14649f360948961  guix-build-e096f5d456f7/output/x86_64-apple-darwin/SHA256SUMS.part
fc099ae55db54182430cdcb34bd01d623b2d740af6658ca851e605f38529c2a8  guix-build-e096f5d456f7/output/x86_64-apple-darwin/bitcoin-e096f5d456f7-x86_64-apple-darwin-unsigned.tar.gz
6cffaff99ffac45039f98c7c023b5bc6b3dae25912272218d3822a36c1eb227d  guix-build-e096f5d456f7/output/x86_64-apple-darwin/bitcoin-e096f5d456f7-x86_64-apple-darwin-unsigned.zip
ae956bf1fb64ac42cc0f9691876fd2b15dd6281a0d15156f32f490232f3b8ad8  guix-build-e096f5d456f7/output/x86_64-apple-darwin/bitcoin-e096f5d456f7-x86_64-apple-darwin.tar.gz
3c3cb778e12766f652def39a2c01e3ee117a043e8eb6a2e1ceb300ebbc4c6158  guix-build-e096f5d456f7/output/x86_64-linux-gnu/SHA256SUMS.part
ed29ef9a227b65632345ec5f94f72f14feeb624f5f0d11b7066b72b71cebd922  guix-build-e096f5d456f7/output/x86_64-linux-gnu/bitcoin-e096f5d456f7-x86_64-linux-gnu-debug.tar.gz
f013aaf6291e855125775e4788978903c853c5a9694bc00939ba32a30ff59ca7  guix-build-e096f5d456f7/output/x86_64-linux-gnu/bitcoin-e096f5d456f7-x86_64-linux-gnu.tar.gz
d86d7d335f985ceecbe138f113b8e4ac63273075af5ff50e594219dd143d5851  guix-build-e096f5d456f7/output/x86_64-w64-mingw32/SHA256SUMS.part
128c5cc7a837dbb5a17f3be7073205e039075834f21487b1f5b035b69943c45c  guix-build-e096f5d456f7/output/x86_64-w64-mingw32/bitcoin-e096f5d456f7-win64-debug.zip
66040bfa03d4e18c39595dbe8a87503812df91054c68aacbce83926f855ef3b9  guix-build-e096f5d456f7/output/x86_64-w64-mingw32/bitcoin-e096f5d456f7-win64-setup-unsigned.exe
a44d908ba608a534bd63d1bc06ce17806ff4a209406e414c3d20dc1fed96ead4  guix-build-e096f5d456f7/output/x86_64-w64-mingw32/bitcoin-e096f5d456f7-win64-unsigned.tar.gz
f9fa69a3153c6298202c2f896313407e2ff8c837ab1cf4c83a350de7be35374c  guix-build-e096f5d456f7/output/x86_64-w64-mingw32/bitcoin-e096f5d456f7-win64.zip
```

fanquake commented:
Rebased, and fixed an issue with the most recent push. Guix build (aarch64):
```bash
c7aa6bd428ba4ea1c925dafe4df3505ad92e84a196b17c8cb7965d7db231e6a0  guix-build-0388ad0d65b6/output/aarch64-linux-gnu/SHA256SUMS.part
944e734a719886ec26aff12a80be67d28f2c4b7781a341894d3e2ca8477e3497  guix-build-0388ad0d65b6/output/aarch64-linux-gnu/bitcoin-0388ad0d65b6-aarch64-linux-gnu-debug.tar.gz
087ac7a0c1d87a95adacb9fa138aced37a172d4d5199fc75c64831d9f6211972  guix-build-0388ad0d65b6/output/aarch64-linux-gnu/bitcoin-0388ad0d65b6-aarch64-linux-gnu.tar.gz
3b010552293f84e98b715f844fc5d487409bdd9915b32f067109558a329878c3  guix-build-0388ad0d65b6/output/arm-linux-gnueabihf/SHA256SUMS.part
ad923f9de5c2aa379a8332bc84df7273ac55bfb0f34c219064bd740000cd090d  guix-build-0388ad0d65b6/output/arm-linux-gnueabihf/bitcoin-0388ad0d65b6-arm-linux-gnueabihf-debug.tar.gz
393cad9f63d37e18ab523ca6e3730195a8f07d9423f2cd681747099c8964e5d9  guix-build-0388ad0d65b6/output/arm-linux-gnueabihf/bitcoin-0388ad0d65b6-arm-linux-gnueabihf.tar.gz
8f33ab7c9769941dd42d0df1cb4495c42f0af4f66911f4b319a4d75b0ff7a27a  guix-build-0388ad0d65b6/output/arm64-apple-darwin/SHA256SUMS.part
3e8843553961e3835f3b9fcad09200647d486c7d310754563f2efe7ce16edb99  guix-build-0388ad0d65b6/output/arm64-apple-darwin/bitcoin-0388ad0d65b6-arm64-apple-darwin-unsigned.tar.gz
1a38016cdddfd851e1209661bb0254a41c51e928d714bffbfc44215042e9fae7  guix-build-0388ad0d65b6/output/arm64-apple-darwin/bitcoin-0388ad0d65b6-arm64-apple-darwin-unsigned.zip
6b7f955a428cb6defb8b9756ac9c8c074e1b5f2c70b98421e750c59df981ff84  guix-build-0388ad0d65b6/output/arm64-apple-darwin/bitcoin-0388ad0d65b6-arm64-apple-darwin.tar.gz
38fcf966cf2f92b519dba8a9a4ceee884f7d6e928a4bddc008d23b044984766d  guix-build-0388ad0d65b6/output/dist-archive/bitcoin-0388ad0d65b6.tar.gz
92c52e077019b04c14afff325e3eb297e664bb81e80add003639d67d244022ee  guix-build-0388ad0d65b6/output/powerpc64-linux-gnu/SHA256SUMS.part
6d6034d303994d5d67aff1f696ea4b519e1ff2737f8131275a85805473b79e78  guix-build-0388ad0d65b6/output/powerpc64-linux-gnu/bitcoin-0388ad0d65b6-powerpc64-linux-gnu-debug.tar.gz
05a873cf226f5292aedd50e28647e82592d0dc6f3aea43bc1980844523fbf667  guix-build-0388ad0d65b6/output/powerpc64-linux-gnu/bitcoin-0388ad0d65b6-powerpc64-linux-gnu.tar.gz
38ed81ae4fc4d51f62c43ab281892dffc957c63e0b71f8686c580bb3bcfcdc0b  guix-build-0388ad0d65b6/output/riscv64-linux-gnu/SHA256SUMS.part
d9ae89e4b6678310b099c533ba3a0ec7d6e4e67414934653381bf630a2807913  guix-build-0388ad0d65b6/output/riscv64-linux-gnu/bitcoin-0388ad0d65b6-riscv64-linux-gnu-debug.tar.gz
f5b49cac8b1a94e9a4c3837ee9111355530708d2a79112665d6ff845e1d0124a  guix-build-0388ad0d65b6/output/riscv64-linux-gnu/bitcoin-0388ad0d65b6-riscv64-linux-gnu.tar.gz
894dd1107ee9a7c7e9b3a8e872a143f0990046bcc394fab8fc6432cf9e58b2c9  guix-build-0388ad0d65b6/output/x86_64-apple-darwin/SHA256SUMS.part
40872cfb263af083c93be8dfc4c6dd59f087b024dd6dbeaf3b1efbf9b66ea435  guix-build-0388ad0d65b6/output/x86_64-apple-darwin/bitcoin-0388ad0d65b6-x86_64-apple-darwin-unsigned.tar.gz
b52c3a97487f2bc8ae3de2bc3e17339cd2baac39233e8062b03185275464ccc4  guix-build-0388ad0d65b6/output/x86_64-apple-darwin/bitcoin-0388ad0d65b6-x86_64-apple-darwin-unsigned.zip
7455f4851c69d376463fb28a1c82919f7864f26be932898d5861fd6fc9b05e8f  guix-build-0388ad0d65b6/output/x86_64-apple-darwin/bitcoin-0388ad0d65b6-x86_64-apple-darwin.tar.gz
de8f0bb6eb9f933d10eae79f88dd8ad447a6c2fa15270d5d3a6463e5f38de88d  guix-build-0388ad0d65b6/output/x86_64-linux-gnu/SHA256SUMS.part
1362b64d7c51435a99db2f4d57e501502294b76715201a4bc3ad8db988fe3f4f  guix-build-0388ad0d65b6/output/x86_64-linux-gnu/bitcoin-0388ad0d65b6-x86_64-linux-gnu-debug.tar.gz
d710565cf5fe42c7a164e9fd80d031bfe33336801b955b9cdc34a58e9e743f2c  guix-build-0388ad0d65b6/output/x86_64-linux-gnu/bitcoin-0388ad0d65b6-x86_64-linux-gnu.tar.gz
0935d3a6175bb8b8d24a9dbd4acd6265aab1460f9d915aa56a89d9caa9aa0df0  guix-build-0388ad0d65b6/output/x86_64-w64-mingw32/SHA256SUMS.part
6e0a62f6944ab374e04f3133193fd396e3c3424818707b218fcebe8ecb5eb479  guix-build-0388ad0d65b6/output/x86_64-w64-mingw32/bitcoin-0388ad0d65b6-win64-debug.zip
8de57dd9f697f808ae7cabc8564eefb1bfbcfa41a4c7e4c1bd3299b4b4efcf1a  guix-build-0388ad0d65b6/output/x86_64-w64-mingw32/bitcoin-0388ad0d65b6-win64-setup-unsigned.exe
87d4fa7ff52a9b9096021110b2e35a450c925a7eba1ad853c0fb52bd277d51ea  guix-build-0388ad0d65b6/output/x86_64-w64-mingw32/bitcoin-0388ad0d65b6-win64-unsigned.tar.gz
eea8d08ab14394984bc3ade32c515a4371d1fcce3b53ccb85ae004291c6dd977  guix-build-0388ad0d65b6/output/x86_64-w64-mingw32/bitcoin-0388ad0d65b6-win64.zip
```

hebasto commented:
My Guix build:
```
x86_64
c7aa6bd428ba4ea1c925dafe4df3505ad92e84a196b17c8cb7965d7db231e6a0  guix-build-0388ad0d65b6/output/aarch64-linux-gnu/SHA256SUMS.part
944e734a719886ec26aff12a80be67d28f2c4b7781a341894d3e2ca8477e3497  guix-build-0388ad0d65b6/output/aarch64-linux-gnu/bitcoin-0388ad0d65b6-aarch64-linux-gnu-debug.tar.gz
087ac7a0c1d87a95adacb9fa138aced37a172d4d5199fc75c64831d9f6211972  guix-build-0388ad0d65b6/output/aarch64-linux-gnu/bitcoin-0388ad0d65b6-aarch64-linux-gnu.tar.gz
3b010552293f84e98b715f844fc5d487409bdd9915b32f067109558a329878c3  guix-build-0388ad0d65b6/output/arm-linux-gnueabihf/SHA256SUMS.part
ad923f9de5c2aa379a8332bc84df7273ac55bfb0f34c219064bd740000cd090d  guix-build-0388ad0d65b6/output/arm-linux-gnueabihf/bitcoin-0388ad0d65b6-arm-linux-gnueabihf-debug.tar.gz
393cad9f63d37e18ab523ca6e3730195a8f07d9423f2cd681747099c8964e5d9  guix-build-0388ad0d65b6/output/arm-linux-gnueabihf/bitcoin-0388ad0d65b6-arm-linux-gnueabihf.tar.gz
8f33ab7c9769941dd42d0df1cb4495c42f0af4f66911f4b319a4d75b0ff7a27a  guix-build-0388ad0d65b6/output/arm64-apple-darwin/SHA256SUMS.part
3e8843553961e3835f3b9fcad09200647d486c7d310754563f2efe7ce16edb99  guix-build-0388ad0d65b6/output/arm64-apple-darwin/bitcoin-0388ad0d65b6-arm64-apple-darwin-unsigned.tar.gz
1a38016cdddfd851e1209661bb0254a41c51e928d714bffbfc44215042e9fae7  guix-build-0388ad0d65b6/output/arm64-apple-darwin/bitcoin-0388ad0d65b6-arm64-apple-darwin-unsigned.zip
6b7f955a428cb6defb8b9756ac9c8c074e1b5f2c70b98421e750c59df981ff84  guix-build-0388ad0d65b6/output/arm64-apple-darwin/bitcoin-0388ad0d65b6-arm64-apple-darwin.tar.gz
38fcf966cf2f92b519dba8a9a4ceee884f7d6e928a4bddc008d23b044984766d  guix-build-0388ad0d65b6/output/dist-archive/bitcoin-0388ad0d65b6.tar.gz
92c52e077019b04c14afff325e3eb297e664bb81e80add003639d67d244022ee  guix-build-0388ad0d65b6/output/powerpc64-linux-gnu/SHA256SUMS.part
6d6034d303994d5d67aff1f696ea4b519e1ff2737f8131275a85805473b79e78  guix-build-0388ad0d65b6/output/powerpc64-linux-gnu/bitcoin-0388ad0d65b6-powerpc64-linux-gnu-debug.tar.gz
05a873cf226f5292aedd50e28647e82592d0dc6f3aea43bc1980844523fbf667  guix-build-0388ad0d65b6/output/powerpc64-linux-gnu/bitcoin-0388ad0d65b6-powerpc64-linux-gnu.tar.gz
38ed81ae4fc4d51f62c43ab281892dffc957c63e0b71f8686c580bb3bcfcdc0b  guix-build-0388ad0d65b6/output/riscv64-linux-gnu/SHA256SUMS.part
d9ae89e4b6678310b099c533ba3a0ec7d6e4e67414934653381bf630a2807913  guix-build-0388ad0d65b6/output/riscv64-linux-gnu/bitcoin-0388ad0d65b6-riscv64-linux-gnu-debug.tar.gz
f5b49cac8b1a94e9a4c3837ee9111355530708d2a79112665d6ff845e1d0124a  guix-build-0388ad0d65b6/output/riscv64-linux-gnu/bitcoin-0388ad0d65b6-riscv64-linux-gnu.tar.gz
894dd1107ee9a7c7e9b3a8e872a143f0990046bcc394fab8fc6432cf9e58b2c9  guix-build-0388ad0d65b6/output/x86_64-apple-darwin/SHA256SUMS.part
40872cfb263af083c93be8dfc4c6dd59f087b024dd6dbeaf3b1efbf9b66ea435  guix-build-0388ad0d65b6/output/x86_64-apple-darwin/bitcoin-0388ad0d65b6-x86_64-apple-darwin-unsigned.tar.gz
b52c3a97487f2bc8ae3de2bc3e17339cd2baac39233e8062b03185275464ccc4  guix-build-0388ad0d65b6/output/x86_64-apple-darwin/bitcoin-0388ad0d65b6-x86_64-apple-darwin-unsigned.zip
7455f4851c69d376463fb28a1c82919f7864f26be932898d5861fd6fc9b05e8f  guix-build-0388ad0d65b6/output/x86_64-apple-darwin/bitcoin-0388ad0d65b6-x86_64-apple-darwin.tar.gz
de8f0bb6eb9f933d10eae79f88dd8ad447a6c2fa15270d5d3a6463e5f38de88d  guix-build-0388ad0d65b6/output/x86_64-linux-gnu/SHA256SUMS.part
1362b64d7c51435a99db2f4d57e501502294b76715201a4bc3ad8db988fe3f4f  guix-build-0388ad0d65b6/output/x86_64-linux-gnu/bitcoin-0388ad0d65b6-x86_64-linux-gnu-debug.tar.gz
d710565cf5fe42c7a164e9fd80d031bfe33336801b955b9cdc34a58e9e743f2c  guix-build-0388ad0d65b6/output/x86_64-linux-gnu/bitcoin-0388ad0d65b6-x86_64-linux-gnu.tar.gz
0935d3a6175bb8b8d24a9dbd4acd6265aab1460f9d915aa56a89d9caa9aa0df0  guix-build-0388ad0d65b6/output/x86_64-w64-mingw32/SHA256SUMS.part
6e0a62f6944ab374e04f3133193fd396e3c3424818707b218fcebe8ecb5eb479  guix-build-0388ad0d65b6/output/x86_64-w64-mingw32/bitcoin-0388ad0d65b6-win64-debug.zip
8de57dd9f697f808ae7cabc8564eefb1bfbcfa41a4c7e4c1bd3299b4b4efcf1a  guix-build-0388ad0d65b6/output/x86_64-w64-mingw32/bitcoin-0388ad0d65b6-win64-setup-unsigned.exe
87d4fa7ff52a9b9096021110b2e35a450c925a7eba1ad853c0fb52bd277d51ea  guix-build-0388ad0d65b6/output/x86_64-w64-mingw32/bitcoin-0388ad0d65b6-win64-unsigned.tar.gz
eea8d08ab14394984bc3ade32c515a4371d1fcce3b53ccb85ae004291c6dd977  guix-build-0388ad0d65b6/output/x86_64-w64-mingw32/bitcoin-0388ad0d65b6-win64.zip
```

theuni commented:
Looking at the mingw .pc generated by this PR:
```
Libs: -L${libdir} -lzmq
Libs.private:
Requires.private:
```

It looks like we'll need to take https://github.com/zeromq/libzmq/pull/4706 as well for CMake. That can be done as a follow-up though, as it's not yet merged upstream.

hebasto commented on depends/packages/zeromq.mk:
```diff
@@ -3,32 +3,42 @@ $(package)_version=4.3.5
+  $(package)_config_opts += -DWITH_LIBBSD=OFF -DENABLE_CURVE=OFF -DENABLE_CPACK=OFF
+  $(package)_config_opts += -DBUILD_SHARED=OFF -DBUILD_TESTS=OFF -DZMQ_BUILD_TESTS=OFF
+  $(package)_config_opts += -DENABLE_DRAFTS=OFF -DZMQ_BUILD_TESTS=OFF
+  $(package)_cxxflags += -ffile-prefix-map=$($(package)_extract_dir)=/usr
```
I'm curious why this flag became necessary during the migration to CMake to achieve cross-arch reproducibility?

fanquake replied:
To remove the non-deterministic file prefixes being added to the library.

hebasto replied:
Right. I understand what `-ffile-prefix-map=...` does.

My question was about a change that introduced such non-determinism, given that the source code is the same.

The `-ffile-prefix-map=...` flag is neither required when building `zeromq` with Autotools nor used by any other package in depends.

fanquake replied:
> My question was about a change that introduced such non-determinism, given that the source code is the same.

The change is CMake. It's passing the full/path/to/the/source/files during compilation, rather than something like src/file which autotools does, which in turn causes the `__FILE__` macros in the zmq error handling code expand to non-deterministic paths.

hebasto reviewed (commented):
CMake compiles 7 fewer source files compared to Autotools. It skips::
- `gssapi_client.cpp`
- `gssapi_mechanism_base.cpp`
- `gssapi_server.cpp`
- `vmci_address.cpp`
- `vmci_connecter.cpp`
- `vmci.cpp`
- `vmci_listener.cpp`

fanquake commented:
> CMake compiles 7 fewer source files compared to Autotools. It skips::

That's expected. We aren't opting in to either of these features.

hebasto reviewed (approved):
ACK 0388ad0d65b6c9ee802ca641eb01d69fcdd5605d.

> > CMake compiles 7 fewer source files compared to Autotools. It skips::
> 
> That's expected. We aren't opting in to either of these features.

Right. Perhaps, it should be considered an upstream bug in `configure.ac`, as we aren't opting into these features in Autotools either.

fanquake commented:
> Perhaps, it should be considered an upstream bug in configure.ac

The files are compiled unconditionally, and the content is control by defines. If you look at the object files in depends. None have any symbols, so I'm not sure how it's a bug.

hebasto commented:
> Looking at the mingw .pc generated by this PR:
> 
> ```
> Libs: -L${libdir} -lzmq
> Libs.private:
> Requires.private:
> ```
> 
> It looks like we'll need to take [zeromq/libzmq#4706](https://github.com/zeromq/libzmq/pull/4706) as well for CMake. That can be done as a follow-up though, as it's not yet merged upstream.

Done in https://github.com/bitcoin/bitcoin/pull/30508.

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/278.

hebasto commented:
This PR fails to compile on OpenBSD 7.5 :
```
[ 19%] Building CXX object CMakeFiles/objects.dir/src/io_thread.cpp.o
/home/hebasto/bitcoin/depends/work/build/amd64-unknown-openbsd7.5/zeromq/4.3.5-df5b1b9f936/src/io_thread.cpp:14:22: error: static_cast from 'std::nullptr_t' to 'poller_t::handle_t' (aka 'int') is not allowed
    _mailbox_handle (static_cast<poller_t::handle_t> (NULL))
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
